qubes/storage/__init__.py

```diff
@@ -158,6 +162,58 @@ def __xml__(self):
         config = _sanitize_config(self.config)
         return lxml.etree.Element('volume', **config)
 
+    def ephemeral(self):
+        '''Should this volume be encrypted with an ephemeral key in dom0?
+        '''
+        return not self.snap_on_start and not self.save_on_stop and \
+                self.domain is None and self.rw
+
+    async def start_encrypted(self, name):
+        '''
+        Start a volume encrypted with an ephemeral key.
+        This can be implemented as a coroutine.
+
+        The default implementation of this method uses ``cryptsetup(8)`` with a
+        key taken from ``/dev/urandom``.  This is highly secure and works with
+        any storage pool implementation.  Volume implementations should override
+        this method if they can provide a secure and more efficient
+        implementation.
+        '''
+        assert name.startswith('/dev/mapper/'), 'Invalid path %r passed to cryptsetup' % name
+        async with self._crypt_lock:
+            if self.script is not None:
+                self._not_implemented("starting an encrypted volume with a script")
+            if self.encrypted_name is not None:
+                await qubes.utils.cryptsetup('--', 'close', self.encrypted_name)
+                self.encrypted_name = None
+            await qubes.utils.coro_maybe(self.start())
+            self.encrypted_name = name = name[12:]
```

Comment:
Maybe move the `name = name[12:]` just after the assert, so it's clear what this 12 is about.